### PR TITLE
Cache aware streaming nfa

### PR DESCRIPTION
**Pull request by Slyne**
# What does this PR do ?
Add cache aware streaming model in nemo force aligner

Should be easy for ctc segmentation tool to call and get the log probs as well.

Most of the codes are copied from speech_to_text_cache_aware_streaming_infer.py.

**Collection**: [Note which collection this PR will affect]
ASR

# Changelog 

# Usage
* You can potentially add a usage example below
```
nemo_model="nemo_models/cache_aware_ctc.nemo"
cd NeMo/tools/nemo_forced_aligner/
python3 align.py \
        batch_size=2 \
        model_path=$nemo_model \
        model_downsample_factor=4 \
        manifest_filepath=$manifest \
        output_dir=./streaming_ctm_with_pred \
        align_using_pred_text=true \
        simulate_cache_aware_streaming=true
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/mixins/mixins.py [hunk 1]
```diff
@@ -526,13 +531,153 @@ def conformer_stream_step(
...
-        )
+        ]
+        if return_log_probs:
+            result.append(log_probs)
+            result.append(encoded_len)
+
+        return tuple(result)
+
+    @torch.no_grad()
+    def transcribe_simulate_cache_aware_streaming(
```
Comment (VahidooX) [inline nemo/collections/asr/parts/mixins/mixins.py:548]: If you just want the final transcriptions, you don't need this method and no need to simulate streaming. If you call the regular transcribe method, it gives you the exact same result as this method. When you have the whole audio, the result of offline and streaming is exactly the same for cache-aware models.

File: tools/nemo_forced_aligner/align.py [hunk 2]
```diff
@@ -51,6 +51,8 @@
         If the ASR model is a QuartzNet model, its downsample factor is 2.
         If the ASR model is a Conformer CTC model, its downsample factor is 4.
         If the ASR model is a Citirnet model, its downsample factor is 8.
+    simulate_cache_aware_streaming: only valid for cache aware streaming conformer model, set true to do somulate
```
Comment (VahidooX) [inline tools/nemo_forced_aligner/align.py:54]: Is there any docs on these aligners?

File: tools/nemo_forced_aligner/utils/data_prep.py [hunk 3]
```diff
@@ -299,9 +298,14 @@ def get_batch_tensors_and_boundary_info(manifest_lines_batch, model, separator,
...
     audio_filepaths_batch = [line["audio_filepath"] for line in manifest_lines_batch]
     B = len(audio_filepaths_batch)
-    with torch.no_grad():
-        hypotheses = model.transcribe(audio_filepaths_batch, return_hypotheses=True, batch_size=B)
-
+    if not simulate_cache_aware_streaming:
+        with torch.no_grad():
+            hypotheses = model.transcribe(audio_filepaths_batch, return_hypotheses=True, batch_size=B)
+    else:
+        with torch.no_grad():
```
Comment (VahidooX) [inline tools/nemo_forced_aligner/utils/data_prep.py:305]: @titu1994 what is the safest way to use a model in inference mode? is torch.nograd() enough?

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/mixins/mixins.py (hunk 1 above)
Comment (Slyne) [reply nemo/collections/asr/parts/mixins/mixins.py:548]: I need the log_probs. If I just use the regular transcribe() method, it will cause GPU OOM issue for long audios.

File: nemo/collections/asr/parts/mixins/mixins.py [hunk 4]
```diff
@@ -453,13 +454,17 @@ def conformer_stream_step(
...
+            return_log_probs: whether to return the log probs, only valid for ctc model
 
         Returns:
             greedy_predictions: the greedy predictions from the decoder
             all_hyp_or_transcribed_texts: the decoder hypotheses for Transducer models and the transcriptions for CTC models
             cache_last_channel_next: the updated tensor cache for last channel layers to be used for next streaming step
             cache_last_time_next: the updated tensor cache for last time layers to be used for next streaming step
             best_hyp: the best hypotheses for the Transducer models
+
+            log_probs: the logits tensor of current streaming chunk, only returned when return_log_probs=True
```
Comment (VahidooX) [inline nemo/collections/asr/parts/mixins/mixins.py:466]: Let's check if the model is CTC when log_probs is enabled.

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: tools/nemo_forced_aligner/utils/data_prep.py [hunk 5]
```diff
@@ -299,9 +298,14 @@ def get_batch_tensors_and_boundary_info(manifest_lines_batch, model, separator,
     # and (optionally) the predicted ASR text from the hypotheses
     audio_filepaths_batch = [line["audio_filepath"] for line in manifest_lines_batch]
     B = len(audio_filepaths_batch)
-    with torch.no_grad():
-        hypotheses = model.transcribe(audio_filepaths_batch, return_hypotheses=True, batch_size=B)
-
+    if not simulate_cache_aware_streaming:
+        with torch.no_grad():
+            hypotheses = model.transcribe(audio_filepaths_batch, return_hypotheses=True, batch_size=B)
```
[the remark this answers is not in the document]
Comment (Slyne) [reply tools/nemo_forced_aligner/utils/data_prep.py:303]: The way of switching to local attention by sacrifice a little accuracy seems interesting. But for now, let's only consider the arbitrary cache aware streaming model.

File: tools/nemo_forced_aligner/align.py (hunk 2 above)
Comment (VahidooX) [reply tools/nemo_forced_aligner/align.py:54]: Has the docs updated with this new feature and how users may use it?
Is there any plan to add these docs to the main docs of nemo?

File: tools/nemo_forced_aligner/utils/data_prep.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (VahidooX) [reply tools/nemo_forced_aligner/utils/data_prep.py:303]: Sure, let's keep the cache-aware but as we don't we have any cache-aware checkpoints publicly available, it would be nice to give users the option of using buffered streaming with offline models with local attention to support long audios. It can be done later if you see useful.

**Review by VahidooX — APPROVED** [review]
LGTM! Just left some minor comments on the docs.

**Review by VahidooX — APPROVED** [review]
